python-package/xgboost/spark/data.py

```diff
@@ -1,192 +1,166 @@
-            f"Rows contain different feature dimensions: Expecting {expected_dims}, got {num_dims}."
-        )
-    return expected_dims
-
-
-def _row_tuple_list_to_feature_matrix_y_w(
-    data_iterator,
-    train,
-    has_weight,
-    has_fit_base_margin,
-    has_predict_base_margin,
-    has_validation: bool = False,
-):
-    """
-    Construct a feature matrix in ndarray format, label array y and weight array w
-    from the row_tuple_list.
-    If train == False, y and w will be None.
-    If has_weight == False, w will be None.
-    If has_base_margin == False, b_m will be None.
-    Note: the row_tuple_list will be cleared during
-    executing for reducing peak memory consumption
-    """
-    # pylint: disable=too-many-locals
-    expected_feature_dims = None
-    label_list, weight_list, base_margin_list = [], [], []
-    label_val_list, weight_val_list, base_margin_val_list = [], [], []
-    values_list, values_val_list = [], []
-
-    # Process rows
-    for pdf in data_iterator:
-        if len(pdf) == 0:
-            continue
-        if train and has_validation:
-            pdf_val = pdf.loc[pdf["validationIndicator"], :]
-            pdf = pdf.loc[~pdf["validationIndicator"], :]
-
-        num_feature_dims = len(pdf["values"].values[0])
-
-        expected_feature_dims = _check_feature_dims(
-            num_feature_dims, expected_feature_dims
-        )
+from xgboost import DMatrix, DataIter, DeviceQuantileDMatrix
+from xgboost.compat import concat, lazy_isinstance
+
+
+def stack_df(df: pd.DataFrame) -> np.ndarray:
+    array = df.values
+    if array.ndim == 1:
+        array = array.reshape(array.shape[0], 1)
+    array = np.stack(array[:, 0])
+    return array
+
+
+def concat_or_none(seq: Optional[Sequence[np.ndarray]]) -> Optional[np.ndarray]:
+    if seq:
+        return concat(seq)
+    return None
+
+
+Alias = namedtuple("Alias", ("data", "label", "weight", "margin", "valid"))
+alias = Alias("values", "label", "weight", "baseMargin", "validationIndicator")
 
-        # Note: each element in `pdf["values"]` is an numpy array.
-        values_list.append(pdf["values"].to_list())
-        if train:
-            label_list.append(pdf["label"].to_numpy())
-        if has_weight:
-            weight_list.append(pdf["weight"].to_numpy())
-        if has_fit_base_margin or has_predict_base_margin:
-            base_margin_list.append(pdf["baseMargin"].to_numpy())
+
+def create_dmatrix_from_partitions(
+    iterator: Iterator[pd.DataFrame], has_validation: bool
+) -> Tuple[DMatrix, Optional[DMatrix]]:
+    train_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+    valid_data: Dict[str, List[np.ndarray]] = defaultdict(list)
+    n_features: List[int] = [0]
+
+    def append(part: pd.DataFrame, name: str, is_valid: bool) -> None:
+        if name in part.columns:
+            array = part[name]
+            if name == alias.data:
+                array = stack_df(array)
+                if n_features[0] == 0:
```

Comment:
we can use `nonlocal n_features` in append function,
so that `n_features` in outer scope can be a `int` type and inside `append` we can still modify it.